sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -18,21 +18,28 @@
 package org.apache.spark.sql.catalyst.expressions
 
 import org.apache.spark.sql.catalyst.analysis.UnresolvedException
-import org.apache.spark.sql.catalyst.errors.TreeNodeException
 import org.apache.spark.sql.types._
 
 case class UnaryMinus(child: Expression) extends UnaryExpression {
   type EvaluatedType = Any
 
+  override lazy val resolved = child.resolved &&
+    (child.dataType.isInstanceOf[NumericType] || child.dataType.isInstanceOf[NullType])
+
   def dataType = child.dataType
   override def foldable = child.foldable
   def nullable = child.nullable
   override def toString = s"-$child"
 
-  lazy val numeric = dataType match {
-    case n: NumericType => n.numeric.asInstanceOf[Numeric[Any]]
-    case other => sys.error(s"Type $other does not support numeric operations")
-  }
+  val numeric =
+    if (resolved) {
+      dataType match {
+        case n: NumericType => n.numeric.asInstanceOf[Numeric[Any]]
+        case n: NullType => UnresolvedNumeric
+      }
+    } else {
+      UnresolvedNumeric
+    }
```

Comment:
`numeric` is only used at runtime, right? If so, our analysis check should catch the unresolved expression (so check if it is resolved for numeric is not needed), right?